Wrote equal_splits_algorith.go and test.

erzel opened the pull request:
@dumbunny Please take a look. 
Note that this code has assertXXX functions. 
As per our discussion in our team-meeting I'll remove them and replace with whatever we will agree on
(either if () {panic()} or if() {return assertionError()} or something else.)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/youtube/vitess/1571)
<!-- Reviewable:end -->

michael-berlin commented:
I like the size of this pull request and the overall structure as well.

In terms of Go readability I left several comments.

---

Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, 46 unresolved discussions, some commit checks failed.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 36 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzGNo7FltcF0bkC2aJ:-KCzGNo7FltcF0bkC2aK:-1944952995) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L36)):*
typo: requires

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 36 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzGkia3mnxIo9sOsFh:-KCzGkia3mnxIo9sOsFi:989217813) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L36)):*
What do you mean with "schema" parameter? I cannot find it in the code?

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 40 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzGrVYQH4QWvTmTZtW:-KCzGrVYQH4QWvTmTZtX:878880619) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L40)):*
Knowing the C++ style guide, I can see that you prefer to have one parameter per line here.

But you also see here that parameter indentation doesn't work so well with the style which gofmt produces :(

It's hard to see when the body of the function starts because it has the same indentation.

Therefore, I'm in favor of keeping things on one line:

```go
func NewEqualSplitsAlgorithm(splitParams *SplitParams, sqlExecuter SQLExecuter) (*EqualSplitsAlgorithm, error) {
```

or alternatively wrap as late as possible:

```go
func NewEqualSplitsAlgorithm(splitParams *SplitParams, sqlExecuter SQLExecuter) (*EqualSplitsAlgorithm, error) {
```

In general, Go doesn't enforce a line length, so go as you like :) 

The general rationale here is that your variable names should be very short in the first place anyway and given that you don't need to wrap your lines. For example, hardcode Go people would probably write "params" or "p" instead of "splitParams". I'm fine with "splitParams" though and not so strict about it.

re: line limits: The only thing which I always try to keep under 80 characters are comments because they are easy to wrap.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 65 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzILpK-VB7mAT3fssg:-KCzILpK-VB7mAT3fssh:-175668771) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L65)):*
The receiver name mustn't be longer than 1 or 2 characters.

I suggest to use "a" here.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 69 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzQCrKdyXPAZh-Hchx:-KCzQCrLDUPF1-vBqNh_:233050197) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L69)):*
Add a short phrase like this to better bridge to the next sentences:

"We do this because"

"using float64 arithmetic ..."

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 70 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzOq60M7fl9ydpUpjy:-KCzOq60M7fl9ydpUpjz:-1448723303) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L70)):*
typo: arithmetic

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 71 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzOz804qmjmWqvsGEw:-KCzOz804qmjmWqvsGEx:-65113357) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L71)):*
I would assume the -1000 value is the min value usually?

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 76 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzPu7q0zoIdP1PafVT:-KCzPu7rmIimVTiqwuWa:-1976607273) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L76)):*
Which "this" do you mean? big.Float? That one will be more efficient? Maybe also write why you're sticking to big.Rat then :)

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 82 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzQf3LvbexzCNLBvkk:-KCzQf3LvbexzCNLBvkl:655026205) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L82)):*
nit: missing comma after if clause:

If the table is empty, minValue ....

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 85 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzW8nSgSRiiwNpIusI:-KCzW8nTB5nxHZoQWwwW:-855737025) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L85)):*
I feel like panic() is to severe here and returning an error is more suitable.

In fact, we have similar code like this already and it returns errors:

https://github.com/youtube/vitess/blob/master/go/vt/worker/clone_utils.go#L304

Mentioning alainjobart@ to chime in here as well.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 109 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzSFqRum8hQMuHAj3t:-KCzSFqSI-2sEb1WKfWo:-1266663524) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L109)):*
You don't need this helper function:


```go
subIntervalSize := new(big.Rat)
```

Another alternative:

```go
var subIntervalSize = &big.Rat{} 
```

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 111 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzSVjDlGNdvwyz56Ds:-KCzSVjDlGNdvwyz56Dt:-1287207421) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L111)):*
Replace newBigRat() with new(big.Rat):

```go
subIntervalSize.Quo(subIntervalSize, new(big.Rat).SetInt64(algorithm.splitParams.splitCount))
```

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 112 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzT8pbmp-5VpJA-3c8:-KCzT8pbmp-5VpJA-3c9:788915683) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L112)):*
This should be:

var result []tuple

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 115 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzTj3V-UuOPRgXxCQr:-KCzTj3V-UuOPRgXxCQs:700325737) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L115)):*
Do not use the helper function.

Alternatives:

```go
boundary := new(big.Rat)
```

```go
var boundary = &big.Rat{} 
```

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 116 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzShSGSaIGDtIN3v1I:-KCzShSGSaIGDtIN3v1J:1929673809) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L116)):*
Replace newBigRat() with new(big.Rat):

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 124 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzU5x57Es20KvKyJpd:-KCzU5x6YIap_wPuE-Yu:-129889893) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L124)):*
Shorter receiver name: a

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 125 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzUDadD9O_pBl4Gfdp:-KCzUDadD9O_pBl4Gfdq:-1369765717) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L125)):*
As said above, personally I'm not in favor of wrapping the line here.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 126 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzV5lhzERsEOyOAtVB:-KCzV5lhzERsEOyOAtVC:1540585932) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L126)):*
When you use "a" as receiver name, you can have the full call on one line.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 132 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzWPG0iu0WWTGKVO50:-KCzWPG0iu0WWTGKVO51:1848641812) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L132)):*
Same comment as above.

Existing similar code currently returns an error: https://github.com/youtube/vitess/blob/master/go/vt/worker/clone_utils.go#L304

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 152 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzXDYtHyPJbkSHQJ5h:-KCzXDYtHyPJbkSHQJ5i:-471876617) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L152)):*
Again: I think this could be an error.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 154 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzXKBVBMduOLDZ0gkE:-KCzXKBVBMduOLDZ0gkF:1836591836) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L154)):*
Personally, I'm in favor of capitalizing all SQL commands:

SELECT MIN(%v), MAX(%v) FROM %v

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 171 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCz_1pjPwGT5vFm6i5k:-KCz_1pkyJMZoRyzzgrf:-118444254) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L171)):*
Maybe leave a comment here why you don't care about the return value of "exact"?

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 174 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzXh9Tfx1ZFNh82zMq:-KCzXh9Tfx1ZFNh82zMr:952088112) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L174)):*
We definitely shouldn't crash here because this could be a user error e.g. they try to run map reduce on column which is not supported?

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 178 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzXwLKtBFBOkoyXpn2:-KCzXwLKtBFBOkoyXpn3:3225055) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L178)):*
Just pass along the error here instead of panic()ing.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 188 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCz_geZSjX6YJhvH7VG:-KCz_geZSjX6YJhvH7VH:-276455263) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L188)):*
This should be an error IMO.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 210 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzbQMrs-3cB-XT44vP:-KCzbQMrs-3cB-XT44vQ:1905590898) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L210)):*
I think you can leave a comment here that you create an Int first because big.Rat has no uint64 setter methods :) Took me a second to figure out why you did this :)

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 211 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCza3iDLGHhzX7-ktGZ:-KCza3iDLGHhzX7-ktG_:-1011187520) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L211)):*
This can be shortened:

```go
return new(big.Rat).SetInt(big.NewInt(0).SetUint64(value))
```

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 216 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzbaXM3Vq2L9RxtTG3:-KCzbaXM3Vq2L9RxtTG4:-1909357153) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L216)):*
This can be shortened:

```go
return new(big.Rat).SetInt64(value)
```

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 222 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzbqdn5Pu5811CiEE9:-KCzbqdn5Pu5811CiEEA:31819474) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L222)):*
This can be shortened:

```go
return new(big.Rat).SetFloat64(value)
```

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 226 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzbxf0On10FspvZodO:-KCzbxf0On10FspvZodP:1503169629) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L226)):*
As discussed above, please remove this one.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm_test.go, line 8 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzd81ianGTVmf2gBhW:-KCzd81ianGTVmf2gBhX:139012184) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm_test.go#L8)):*
The Go review recommendations aren't very strict about grouping imports e.g. see https://github.com/golang/go/wiki/CodeReviewComments#imports

Most importantly, stand library packages should go in their own, first group.

For the rest, you could put everything into one second group or group by project.

The current form doesn't match with that. sqltypes gets its own package but the splitquery_testing code goes together with gomock?

Therefore I suggest one of these two alternatives:

All together:

```go
"github.com/golang/mock/gomock"
"github.com/youtube/vitess/go/sqltypes"
"github.com/youtube/vitess/go/vt/tabletserver/splitquery/splitquery_testing"
```

Third party dependencies get their own group:

```go
"github.com/golang/mock/gomock"

"github.com/youtube/vitess/go/sqltypes"
"github.com/youtube/vitess/go/vt/tabletserver/splitquery/splitquery_testing"
```

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm_test.go, line 13 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzc8vfMJku_iuHTbiZ:-KCzc8vfMJku_iuHTbi_:-1242718578) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm_test.go#L13)):*
It looks like the linter doesn't complain here that the comment didn't start with the function name.

Nonetheless, for consistency I am still in favor of following the style for the test methods as well e.g.:
```go
// TestMultipleBoundariesInt64Col tests splitting the interval [10,60] into 5 parts.
```

Even better I think it makes more sense when this comment goes into the body of the function. The test method name is already self-explanatory and the comment is more about the specific approach of the test (which internval, how many parts) than the purpose (cover int64 types).

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm_test.go, line 13 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzcveVu3nP2vcD1FoP:-KCzcveWLF5lA-47gQl4:448658773) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm_test.go#L13)):*
nit: missing period

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm_test.go, line 18 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzeisMbfK_D2_WPEPo:-KCzeisMbfK_D2_WPEPp:-407249720) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm_test.go#L18)):*
Preferring UPPERCASE here:

SELECT * FROM test_table WHERE int_col > 5

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm_test.go, line 19 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzeWjwh9Jd9p0Hvwvz:-KCzeWjxqGOG42hZD9-c:1184295494) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm_test.go#L19)):*
Write the comment after the parameter and gofmt will format it correctly:

```go
[]string{"int64_col"} /* bindVariables */,
1000 /* splitColumns */,
```

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm_test.go, line 24 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzevxr6SEGKktXKql1:-KCzevxr6SEGKktXKql2:1662251726) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm_test.go#L24)):*
Just do it?

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm_test.go, line 31 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzf2cWgnBzbEA7QrJV:-KCzf2cWgnBzbEA7QrJW:741805242) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm_test.go#L31)):*
SELECT MIN(int64_col), MAX(int64_col) FROM test_table

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm_test.go, line 54 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzfYtrgLu21mseOSfD:-KCzfYtrgLu21mseOSfE:1951489624) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm_test.go#L54)):*
Go prefers the expressions "got" and "want" to refer to the two results.

Additionally, "got" should always go first in the output. Therefore, the code should look like this:

```go
t.Fatalf("got: %v, want: %v", boundaries, expectedBoundaries)
```

Generally, the variable names are called "got" and "want" as well i.e. you can rename them as follows:

expectedBoundaries => want
boundaries => got

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm_test.go, line 58 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzgeDbg7g2TVdF1-Fa:-KCzgeDcvEl8lFchtW9l:786651691) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm_test.go#L58)):*
See comment above. I would move this comment one line down into the body of the function.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm_test.go, line 63 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzgw7fXNLCLz3NKkRM:-KCzgw7fXNLCLz3NKkRN:817248385) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm_test.go#L63)):*
See my comments above since this is duplicated code.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm_test.go, line 64 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzhE1wbQ21at_W7DFr:-KCzhE1wbQ21at_W7DFs:-928974443) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm_test.go#L64)):*
Since this file has a lot of duplicated code, consider to write it as a table driven test:

https://github.com/golang/go/wiki/TableDrivenTests

Looking at this file, I think it would be worth the effort to convert it.

---

*[go/vt/tabletserver/splitquery/split_params.go, line 46 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCziMtz9SJpsg3NHHA0:-KCziMtz9SJpsg3NHHA1:854628407) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/split_params.go#L46)):*
Same comment as in the other file: Personally, I think one parameter per line doesn't make sense for Go.

---

*[go/vt/tabletserver/splitquery/split_params.go, line 53 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCziERMicgBqFcMf2HQ:-KCziERPsqhDasRZ5UX6:-588880) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/split_params.go#L53)):*
Unwrap these three lines into one.

---

*[go/vt/tabletserver/splitquery/split_params.go, line 87 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzm5tH92o2fgm_YTUw:-KCzm5tH92o2fgm_YTUx:1383392255) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/split_params.go#L87)):*
FYI: This is a good example where a hardcode Go reviewer may prefer shorter names e.g. "sc" or "col".

I am actually totally fine with the current form because it makes sense: 1 splitColumn from the array of splitColumns. But others find this too verbose. Therefore just as an FYI.

---

*[go/vt/tabletserver/splitquery/splitquery_testing/utils.go, line 1 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzl83q1bv6CHgR4IdM:-KCzl83q1bv6CHgR4IdN:81725009) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/splitquery_testing/utils.go#L1)):*
If this file remains test code only AND you'll always use it only within one package, you should make the file part of the Go package and rename it with the suffix _test.go.

This way, the Go build tools will ignore the file (in general they ignore files having "." or "_" in their name) and instead the Go will be only run when go test is run.

---

*[go/vt/tabletserver/splitquery/splitquery_testing/utils.go, line 12 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzkI47JN_jxNjlX4wJ:-KCzkI485I2C0PeJb2JN:1023520121) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/splitquery_testing/utils.go#L12)):*
nit: missing period.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/youtube/vitess/1571#-:-KCzlp1bKiGhGUFLQO3P:-1678064535)*
<!-- Sent from Reviewable.io -->

michael-berlin commented:
Regarding Go readability: These slides are nice and contain a lot of tips: https://talks.golang.org/2014/readability.slide

---

Review status: all files reviewed at latest revision, 46 unresolved discussions, some commit checks failed.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 85 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzW8nSgSRiiwNpIusI:-KCztRwTiLc8oUspZUK_:-2056910069) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L85)):*
@alainjobart instead of alainjobart@ to trigger him.

Github is hard.

---

*[go/vt/tabletserver/splitquery/splitquery_testing/utils.go, line 1 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzl83q1bv6CHgR4IdM:-KCztFSvrHKBHAzzTAbt:423096082) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/splitquery_testing/utils.go#L1)):*
To clarify things: I meant this comment only for this file. Even if you can move this file out of this package, you probably still want to have this separate package for the generated mock file. I think that's fine.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/youtube/vitess/1571#-:-KCztaFjLRoT5KUNf7Ls:863485900)*
<!-- Sent from Reviewable.io -->

erzel commented:
Review status: all files reviewed at latest revision, 46 unresolved discussions, some commit checks failed.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 85 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzW8nSgSRiiwNpIusI:-KD-RlP_Yr-Z6vNDmPal:-293745587) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L85)):*
How can this happen if it's not a programming bug? Either  the split_column is empty (contains only NULLs)  and then both max and min will be NULLs or it has non-NULL values in which case both the minimum and maximum be non-NULL. This sounds like a logic violation to me, but it's not a big-deal. If you insist I'll make it an error.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/youtube/vitess/1571)*
<!-- Sent from Reviewable.io -->

michael-berlin commented:
Review status: all files reviewed at latest revision, 46 unresolved discussions, some commit checks failed.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 85 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzW8nSgSRiiwNpIusI:-KD-_WsIrvjOLi8QdFjs:929064398) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L85)):*
I agree with you that this is a programming error.

But nonetheless, we used errors for this type of error as well :)

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/youtube/vitess/1571)*
<!-- Sent from Reviewable.io -->

erzel commented:
Thank a lot Michael for the careful look and the comments and the links to more go resources: I'm learning a lot :-). Answer's inline. I haven't yet refactored the equal_splits_algorithm_test .go as you suggested. In the meantime you can take a look at my responses.

---

Review status: all files reviewed at latest revision, 46 unresolved discussions, some commit checks failed.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 36 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzGNo7FltcF0bkC2aJ:-KD08pRb8XioVNasD5zR:-319227405) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L36)):*
Removed this line entirely. It's anachronistic. Thanks for catching. 

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 40 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzGrVYQH4QWvTmTZtW:-KD09c7NWITKt_PKvnDh:-1355651569) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L40)):*
I use a 100-char limit and prefer not to have my lines wrapped beyond this. I also prefer long(er) parameters names than just a single letter and so would
have to encounter this issue in the future. I have two suggestions:
1. A blank line after 'func' -- it seems gofmt leaves it alone.
func NewEqualSplitsAlgorithm(
     splitParams *SplitParams, sqlExecuter SQLExecuter) (*EqualSplitsAlgorithm, error) {

	assertEqual(len(splitParams.splitColumns), len(splitParams.splitColumnTypes))
	...
}

2. Wrapping  later:
func NewEqualSplitsAlgorithm(splitParams *SplitParams, sqlExecuter SQLExecuter) (
	*EqualSplitsAlgorithm, error) {
	assertEqual(len(splitParams.splitColumns), len(splitParams.splitColumnTypes))
        ...
}

What do you think?

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 65 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzILpK-VB7mAT3fssg:-KD0Biq5UkT69sZYfc7b:-499146009) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L65)):*
Done.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 69 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzQCrKdyXPAZh-Hchx:-KD0LARnGjt-hUqmgmoh:-499146009) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L69)):*
Done.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 70 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzOq60M7fl9ydpUpjy:-KD0CdUZN5AuEd0X3VoD:-499146009) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L70)):*
Done.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 71 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzOz804qmjmWqvsGEw:-KD0CanAIDGnI7k4Bmv6:1711695259) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L71)):*
Yes :-). Thanks.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 76 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzPu7q0zoIdP1PafVT:-KD0ETviFRMdTJeEfO57:-1058710114) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L76)):*
Changed. PTAL.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 82 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzQf3LvbexzCNLBvkk:-KD0F0lZPVM8KQO2lnhw:-499146009) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L82)):*
Done.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 85 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzW8nSgSRiiwNpIusI:-KD0F4ms45I7t49uaD37:-2097701512) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L85)):*
I'd argue that those places need to change (but it's of low priority). 

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 109 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzSFqRum8hQMuHAj3t:-KD0Go6v2vOsB5-4DbJC:-1939532825) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L109)):*
Done. I didn't know about 'new'. Thanks. 

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 111 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzSVjDlGNdvwyz56Ds:-KD0Guqk3pE8RYrIn03y:-499146009) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L111)):*
Done.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 112 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzT8pbmp-5VpJA-3c8:-KD0H431eZPLy4f9evoB:-499146009) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L112)):*
Done.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 115 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzTj3V-UuOPRgXxCQr:-KD0H6_mvVtWgPNovZHV:-499146009) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L115)):*
Done.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 116 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzShSGSaIGDtIN3v1I:-KD0H8IssjlPIqEq7ZXP:-499146009) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L116)):*
Done.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 124 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzU5x57Es20KvKyJpd:-KD0HCI0tQGr4jbSQcIw:-499146009) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L124)):*
Done.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 125 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzUDadD9O_pBl4Gfdp:-KD0HK_os2PJFYHHsPAH:1783069918) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L125)):*
When I use 'a' I don't need to wrap. So I changed it.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 126 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzV5lhzERsEOyOAtVB:-KD0HV9DnSt8eij0WyQj:-499146009) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L126)):*
Done.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 132 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzWPG0iu0WWTGKVO50:-KD0mTaTh_yfFyqiIp1u:-164630557) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L132)):*
See my reply to your comment in line 188. I've sent a query that should only return 1 row. If that's violated there is not much I can do but bail out to our RPC boundary.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 152 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzXDYtHyPJbkSHQJ5h:-KD0LDWqGz7Mq36rVqRi:-1998771596) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L152)):*
The SplitParams constructor (see line 72 in split_params.go) already checks this and returns an error if it's violated. Here it's a programming error.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 154 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzXKBVBMduOLDZ0gkE:-KD0HzO3rpHuNhSBVC4b:-887863794) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L154)):*
Here it seems like the sqlparser formats  SQL's as lowercase:
https://github.com/youtube/vitess/blob/master/go/vt/sqlparser/ast.go#L193

but the logs show them in upper-case. So I'm a little confused as to what the convention should be.
Adding @sougou for clarification.

I'd like to be consistent here.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 171 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCz_1pjPwGT5vFm6i5k:-KD0K38q2sElm0TBSkoy:-499146009) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L171)):*
Done.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 174 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzXh9Tfx1ZFNh82zMq:-KD0K6MNA4xQfcNCUmV9:301866166) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L174)):*
This is already checked (and an error returned) in 'NewEqualSplitsAlgorithm'. So here, it's a programming error.  See line 47.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 178 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzXwLKtBFBOkoyXpn2:-KD0LcK4mEuncxqr7Nrm:-82059650) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L178)):*
After our team discussion I hope you agree that this is clearly a programming error.  ValueFromBytes should not fail for a big.Rat that is written into a string.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 188 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCz_geZSjX6YJhvH7VG:-KD0NDy3LPsulJwVtcEw:-1371365443) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L188)):*
I've added an error here; however in the calling code I'm panicking if I get an error returned from here.
I've sent a 'SELECT min(x), max(x) FROM table' SQL statement, where 'x' has numeric type and then try to parse the result as numeric . If this doesn't succeed there is something wrong with my program logic (or a bug in MySQL). I think an assertion with a stack-trace is more appropriate. I don't think an error will tell the user much anyway.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 210 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzbQMrs-3cB-XT44vP:-KD0KiEYvaG-CzpIi6bB:-499146009) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L210)):*
Done.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 211 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCza3iDLGHhzX7-ktGZ:-KD0L-H_czUEelm25IlX:-499146009) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L211)):*
Done.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 216 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzbaXM3Vq2L9RxtTG3:-KD0L04fFDaVF59OmKo2:-499146009) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L216)):*
Done.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 222 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzbqdn5Pu5811CiEE9:-KD0L0fOrfEnAR1AFR7x:-499146009) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L222)):*
Done.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 226 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzbxf0On10FspvZodO:-KD0L1_mwR0lagBTAkOe:-499146009) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L226)):*
Done.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm_test.go, line 8 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzd81ianGTVmf2gBhW:-KD0fHzJS3PqAtPSl-RQ:1706342951) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm_test.go#L8)):*
Done..

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm_test.go, line 13 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzc8vfMJku_iuHTbiZ:-KD0RKBI1nkain6cRcSo:-499146009) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm_test.go#L13)):*
Done.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm_test.go, line 13 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzcveVu3nP2vcD1FoP:-KD0fNz1A9FTrtZwEOE9:1706342951) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm_test.go#L13)):*
Done..

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm_test.go, line 18 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzeisMbfK_D2_WPEPo:-KD0fmbpaUwishFlNg8H:-17822727) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm_test.go#L18)):*
See reply above.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm_test.go, line 19 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzeWjwh9Jd9p0Hvwvz:-KD0g1VfrDm28lSptV1p:380124153) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm_test.go#L19)):*
I'm not sure I understand. When I swap the comment and the commas gofmt swaps them back.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm_test.go, line 24 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzevxr6SEGKktXKql1:-KD0dZPKbE3HpMc0TF7r:1274154130) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm_test.go#L24)):*
I already have in my next PR. Would you mind if we submit as is?

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm_test.go, line 31 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzf2cWgnBzbEA7QrJV:-KD0dg2BGoIxn7quEf8O:-639453191) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm_test.go#L31)):*
See my previous reply to this. Waiting to see what the convention is.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm_test.go, line 54 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzfYtrgLu21mseOSfD:-KD0dnbWB8uRNvTeLqW_:-1060341830) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm_test.go#L54)):*
Sure. Is there somewhere I can read more about this convention?

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm_test.go, line 58 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzgeDbg7g2TVdF1-Fa:-KD0dt5zh5lcYGS_26Yl:-499146009) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm_test.go#L58)):*
Done.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm_test.go, line 64 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzhE1wbQ21at_W7DFr:-KD0eirHKJSVE5wPrCX4:19222242) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm_test.go#L64)):*
Ok. Let me do that.

---

*[go/vt/tabletserver/splitquery/split_params.go, line 46 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCziMtz9SJpsg3NHHA0:-KD0eY5sgZtCHUTkS9WA:-499146009) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/split_params.go#L46)):*
Done.

---

*[go/vt/tabletserver/splitquery/split_params.go, line 53 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCziERMicgBqFcMf2HQ:-KD0eZGK6yfPaXKWeeYP:-499146009) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/split_params.go#L53)):*
Done.

---

*[go/vt/tabletserver/splitquery/split_params.go, line 87 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzm5tH92o2fgm_YTUw:-KD0e_ix8Egy3VVx4Qfy:-939511033) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/split_params.go#L87)):*
Ack. Let's keep this between us then :-).

---

*[go/vt/tabletserver/splitquery/splitquery_testing/utils.go, line 1 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzl83q1bv6CHgR4IdM:-KD0cHcFhXN2Q_nY7DTH:-411068415) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/splitquery_testing/utils.go#L1)):*
Done. Although it seems 'go build' only ignore files that start with '.' or '_'. And for these  'go test' also ignores them. 
So I had to call my file testutils_test.go which is a little confusing. Do you know if there's a way I can easily specify which files to only include when building a test?

---

*[go/vt/tabletserver/splitquery/splitquery_testing/utils.go, line 12 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzkI47JN_jxNjlX4wJ:-KD0RRsLs6s94kPpGFAa:-499146009) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/splitquery_testing/utils.go#L12)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/youtube/vitess/1571#-:-KD0n-Kx1ahybR9yBlB-:1153002261)*
<!-- Sent from Reviewable.io -->

michael-berlin commented:
LGTM

Feel free to merge this PR after you addressed the remaining comments.

---

Reviewed 7 of 7 files at r2.
Review status: all files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 40 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzGrVYQH4QWvTmTZtW:-KCzOG3_g5iq750QbjIO:-107034321) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L40)):*
In the "wrapping later example" the whole function signature is currently on one line. You probably meant something else here?

In general: you can use three backticks as start and end marker to mark a code block. In addition, you can tell Github which syntax highlighting to use.

See: https://help.github.com/articles/basic-writing-and-formatting-syntax/#quoting-code

re: the current form. The new line works for me. But having no new line there would probably be fine with me as well.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 85 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzW8nSgSRiiwNpIusI:-KD10fmeCWoM9oFMaih8:1020329392) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L85)):*
Keeping the panic because we consider this condition an assertion is fine with me. Therefore, I'm resolving this comment.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 125 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzUDadD9O_pBl4Gfdp:-KD11oAQbtO0yziKaf4Q:1099103656) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L125)):*
Hooray :)

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 132 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzWPG0iu0WWTGKVO50:-KD11x0k8xnGbAe86ouE:175262642) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L132)):*
Keeping panic() is fine with me given our latest discussion.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 152 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzXDYtHyPJbkSHQJ5h:-KD12geBILJXWjLOapvF:1374620311) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L152)):*
Keeping panic() as assertion is fine with me then.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 174 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzXh9Tfx1ZFNh82zMq:-KD13TVxlrcb_y7MZGBn:1374620311) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L174)):*
Keeping panic() as assertion is fine with me then.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 178 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzXwLKtBFBOkoyXpn2:-KD13XtiIXirF_6hCsUE:1374620311) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L178)):*
Keeping panic() as assertion is fine with me then.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm_test.go, line 19 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzeWjwh9Jd9p0Hvwvz:-KD18_ljBGlQppiLK4uP:-495111964) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm_test.go#L19)):*
You are right :(

It doesn't swap it back when you put the comment in front of the variable:

```go
/* bindVariables */ []string{"int64_col"},
 /* splitColumns */ 1000,
```

Feel free to change it if you want to. Keeping it is otherwise fine with me as well.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm_test.go, line 24 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzevxr6SEGKktXKql1:-KD14YpvJ_Hzuam3eP90:-1179112826) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm_test.go#L24)):*
No, submitting as is is fine.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm_test.go, line 54 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzfYtrgLu21mseOSfD:-KD155sPOyNIWlZJ_iZB:-1092690037) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm_test.go#L54)):*
https://github.com/golang/go/wiki/CodeReviewComments#useful-test-failures

---

*[go/vt/tabletserver/splitquery/testutils_test.go, line 1 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzl83q1bv6CHgR4IdM:-KD16-3CIg8Ubv_1KlJ_:-1155621636) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/testutils_test.go#L1)):*
I think it's better to keep things simple and keep _test.go as suffix.

In general, you can use Go build tags to exclude a file from getting build e.g. for specific platforms: https://golang.org/pkg/go/build/#hdr-Build_Constraints

But it looks like you cannot tell go build to include a file only for tests?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/youtube/vitess/1571#-:-KD17rK8FgWIESoP6zBZ:743889145)*
<!-- Sent from Reviewable.io -->

erzel commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 132 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzWPG0iu0WWTGKVO50:-KD1Oz-u__nX5C_q3VOV:1710950521) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L132)):*
Cool. Actually, thinking this over on my way home I have to say I'm not as convinced as I was about whether this (and other violations from an external system--MySQL in this case) should be an error or an assertion. I'm still leaning to an assertion for this case, but I think sometimes the line can be fairly thin. We can leave as it is for now, but I'll also be fine with changing to match the behavior of our code elsewhere.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/youtube/vitess/1571)*
<!-- Sent from Reviewable.io -->

dumbunny commented on go/vt/tabletserver/splitquery/equal_splits_algorithm.go:
```diff
@@ -0,0 +1,226 @@
+	// subIntervalSize = (max - min) / a.splitParams.splitCount
+	subIntervalSize := new(big.Rat)
+	subIntervalSize.Sub(max, min)
+	subIntervalSize.Quo(subIntervalSize, new(big.Rat).SetInt64(a.splitParams.splitCount))
```
Better if boundary := min at the start of the loop, then
boundary.Add(boundary, subIntervalSize) every pass through the loop.

erzel commented:
Review status: 6 of 8 files reviewed at latest revision, 8 unresolved discussions.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 40 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzGrVYQH4QWvTmTZtW:-KDAoQu3K7uqRlhK_c33:-499146009) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L40)):*
Done.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 154 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCzXKBVBMduOLDZ0gkE:-KDArCsYAt7bQCdeczIF:-1488901824) ([raw file](https://github.com/youtube/vitess/blob/5d67fb4b33d43a0e753dcacaf20972fd47b8a098/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L154)):*
Looks like Sugu's recommendation is to use what the parser generates. I'm going to stick with that for now. Michael, if you object, I can change in a later PR.

---

*[go/vt/tabletserver/splitquery/equal_splits_algorithm.go, line 117 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1571#-KCwPMFk8r3TDUayXU0i-r2-117:-KDAmHwdGndWLnSgSOvY:563490447) ([raw file](https://github.com/youtube/vitess/blob/9320321b8dfb3faefa2e35b6536b843306cd4104/go/vt/tabletserver/splitquery/equal_splits_algorithm.go#L117)):*
Right. I initially had this with float64 arithmetic where adding an interval in each iteration would be prone to numerical errors if the interval is small; however for big.Rat this will be correct and likely more efficient. Thanks. Changed.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/youtube/vitess/1571)*
<!-- Sent from Reviewable.io -->
